fuel-core/src/service/genesis.rs

```diff
@@ -30,126 +36,295 @@ use fuel_core_interfaces::{
+            .into_owned();
+        let state_root = self
+            .database
+            .storage::<ContractsState>()
+            .root(&self.contract_id)?;
+        let balance_root = self
+            .database
+            .storage::<ContractsAssets>()
+            .root(&self.contract_id)?;
+
+        let contract_hash = *Hasher::default()
+            // `ContractId` already is based on contract's code and salt so we don't need it.
+            .chain(self.contract_id.as_ref())
+            .chain(utxo.tx_id().as_ref())
+            .chain([utxo.output_index()])
+            .chain(state_root.as_slice())
+            .chain(balance_root.as_slice())
+            .finalize();
+
+        Ok(contract_hash)
+    }
+}
+
+impl Merklization for ConsensusParameters {
+    fn root(&mut self) -> Result<MerkleRoot> {
+        // TODO: Define hash algorithm for `ConsensusParameters`
+        let params_hash = Hasher::default()
+            .chain(bincode::serialize(&self)?)
+            .finalize();
+        Ok(params_hash.into())
+    }
+}
+
+impl Merklization for ChainConfig {
+    fn root(&mut self) -> Result<MerkleRoot> {
+        // TODO: Hash settlement configuration
+        let config_hash = *Hasher::default()
+            // `ContractId` based on contract's code and salt so we don't need it.
+            .chain(self.block_gas_limit.to_be_bytes())
+            .chain(self.transaction_parameters.root()?)
+            .finalize();
+
+        Ok(config_hash)
+    }
+}
+
 impl FuelService {
     /// Loads state from the chain config into database
     pub(crate) fn initialize_state(config: &Config, database: &Database) -> Result<()> {
-        // start a db transaction for bulk-writing
-        let mut import_tx = database.transaction();
-        let database = import_tx.as_mut();
-
         // check if chain is initialized
         if database.get_chain_name()?.is_none() {
-            // initialize the chain id
-            database.init(config)?;
+            // start a db transaction for bulk-writing
+            let mut import_tx = database.transaction();
+            let database = import_tx.as_mut();
 
-            if let Some(initial_state) = &config.chain_conf.initial_state {
-                Self::init_coin_state(database, initial_state)?;
-                Self::init_contracts(database, initial_state)?;
-                Self::init_da_messages(database, initial_state)?;
-            }
-        }
+            Self::add_genesis_block(config, database)?;
 
-        // Write transaction to db
-        import_tx.commit()?;
+            // Write transaction to db
+            import_tx.commit()?;
+        }
 
         Ok(())
     }
 
+    pub fn add_genesis_block(config: &Config, database: &mut Database) -> Result<()> {
+        // Initialize the chain id and height.
+        database.init(&config.chain_conf)?;
+
+        let chain_config_hash = config.chain_conf.clone().root()?.into();
+        let coins_hash =
```

Comment:
Hmm I'm a little torn on this approach of reading a partially setup DB to generate the genesis block.

- On one hand, deriving the genesis block fields from the database ensures that it matches what is actually contained in the state. 
- On the other hand, it seems like deriving the genesis block field could be stateless op done without a db by just inspecting the `chain_conf` values.

The main risk I see with the current approach is that in order to generate a genesis block for comparison or other tools, it requires a fully clean db initialized in a very specific way. This could hamstring our options in the future such as patching existing environments with a missing genesis block, as the DB would already be contaminated with lots of state after genesis. 

It seems like we either run the risk that the genesis block contains more than just genesis state in some unforeseen future usages of the API. Or we run the risk that the computed genesis block doesn't actually align to the initial DB state, however this issue seems less likely to me. IMO it seems like being able to derive the genesis block from the `chain_conf` without a pre-populated DB would be a safer alternative.

Reply:
`Coin`s and `Message`s hashes are based only on data from the `ChainConfig`, so a dirty database will not affect it.
`Contract`s state assumes that the contract didn't exist before, so the state for it also should be clean.

The hashing takes into account only entities added during this initialisation, so dirty database will not break it

Reply:
I added check that inserted values are new

Reply:
This should suffice for now, especially considering the extra work to impl merklizing using only chain_conf data instead of just using the merkle storage interfaces. Also contract UTXO ID's may be tricky.